collect_app/src/main/java/org/odk/collect/android/tasks/InstanceUploaderTask.java

```diff
@@ -1,80 +1,139 @@
-                            .filter(instance -> instance.getStatus().equals(Instance.STATUS_SUBMITTED))
-                            .filter(instance -> InstanceAutoDeleteChecker.shouldInstanceBeDeleted(formsRepository, isFormAutoDeleteOptionEnabled, instance));
+            try {
+                String destinationUrl = uploader.getUrlToSubmitTo(instance, deviceId, completeDestinationUrl, null);
+                String customMessage = uploader.uploadOneSubmission(instance, destinationUrl);
+                outcome.messagesByInstanceId.put(instance.getDbId().toString(),
+                        customMessage != null ? customMessage : getLocalizedString(Collect.getInstance(), org.odk.collect.strings.R.string.success));
 
-                    DeleteInstancesTask dit = new DeleteInstancesTask(instancesRepository, formsRepository);
-                    dit.execute(instancesToDelete.map(Instance::getDbId).toArray(Long[]::new));
-                }
+                Analytics.log(SUBMISSION, "HTTP", Collect.getFormIdentifierHash(instance.getFormId(), instance.getFormVersion()));
+            } catch (FormUploadAuthRequestedException e) {
+                outcome.authRequestingServer = e.getAuthRequestingServer();
+                // Don't add the instance that caused an auth request to the map because we want to
+                // retry. Items present in the map are considered already attempted and won't be
+                // retried.
+            } catch (FormUploadException e) {
+                outcome.messagesByInstanceId.put(instance.getDbId().toString(),
+                        e.getMessage());
             }
         }
+
+        // Delete instances that were successfully sent and that need to be deleted
+        // either because app-level auto-delete is enabled or because the form
+        // specifies it.
+        Set<String> instanceIds = outcome.messagesByInstanceId.keySet();
+
+        boolean isFormAutoDeleteOptionEnabled;
+
+        // The custom configuration from the third party app overrides
+        // the app preferences set for delete after submission
+        if (deleteInstanceAfterSubmission != null) {
+            isFormAutoDeleteOptionEnabled = deleteInstanceAfterSubmission;
+        } else {
+            isFormAutoDeleteOptionEnabled = settingsProvider.getUnprotectedSettings().getBoolean(ProjectKeys.KEY_DELETE_AFTER_SEND);
+        }
+
+        Stream<Instance> instancesToDelete = instanceIds.stream()
+                .map(id -> new InstancesRepositoryProvider(Collect.getInstance()).get().get(Long.parseLong(id)))
+                .filter(instance -> instance.getStatus().equals(Instance.STATUS_SUBMITTED))
+                .filter(instance -> InstanceAutoDeleteChecker.shouldInstanceBeDeleted(formsRepository, isFormAutoDeleteOptionEnabled, instance));
+
+        InstanceDeleter instanceDeleter = new InstanceDeleter(instancesRepository, formsRepository);
```

Comment:
I'd prefer the `InstancesDataService` here, but I'm guessing we end up in a weird state where we've locked the `ChangeLock` so can't then call `deleteInstances`? If so, I guess down the line we can probably have a `InstancesDataService#submitInstances` that covers submission and the clean up. That will probably make sense once we rework the upload UI.

Reply:
> I'd prefer the InstancesDataService here, but I'm guessing we end up in a weird state where we've locked the ChangeLock so can't then call deleteInstances?

Correct.

>If so, I guess down the line we can probably have a InstancesDataService#submitInstances that covers submission and the clean up. That will probably make sense once we rework the upload UI. 

Yeah makes sens we also need to rework the whole task since it's still android AsyncTask. Let's do that in a separate pr.